Adding conda.yaml and MLProject and updating train_predict.py

p2rk2h opened the pull request:
Initial checkin for conda.yaml and MLProject.  Adding mlflow input for 'hidden_units' list input, which converts from string to list in train_predict.py.

p2rk2h committed:
conda.yaml, MLproject

Initial checkin for `mlflow run ...`

p2rk2h committed:
train_predict.py

'hidden_units' to use string input and split( ) to convert into a list (see also MLproject).

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+                    help='hidden units (default: "50 20")')
+parser.add_argument('--steps', type=int, default=1000, metavar='N',
+                    help='steps (default: 1000)')
+parser.add_argument('--tag_constants', type=str, default=tf.saved_model.tag_constants.SERVING, metavar='S',
```
Remove `tag_constants` argument.

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
 
+# Command-line arguments
+parser = argparse.ArgumentParser(description='Tensorflow Example')
+# parser.add_argument('--hidden_units', nargs='*', default='50 20'.split( ), metavar='L',
```
Remove commented code - the uncommented `hidden-units` parameter specification that you included below is better anyway!

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+parser = argparse.ArgumentParser(description='Tensorflow Example')
+# parser.add_argument('--hidden_units', nargs='*', default='50 20'.split( ), metavar='L',
+#                    help='hidden units (default: [50, 20])')  #  https://docs.python.org/3/library/argparse.html#nargs
+parser.add_argument('--hidden_units', type=str, default='50 20', metavar='L',
```
Nit: Can we use a hyphen as a separator in the parameter name instead of an underscore? `hidden_units` --> `hidden-units`.

dbczumar replied:
Remove `metavar` from the parameter definition.

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+# parser.add_argument('--hidden_units', nargs='*', default='50 20'.split( ), metavar='L',
+#                    help='hidden units (default: [50, 20])')  #  https://docs.python.org/3/library/argparse.html#nargs
+parser.add_argument('--hidden_units', type=str, default='50 20', metavar='L',
+                    help='hidden units (default: "50 20")')
```
Nit: Instead of copying the default "50 20" value, use `%(default)s`. Let's also make the help text a little more descriptive. The changed line will look like

```
help='The number of hidden units in each of the two hidden unit layers (default: %(default)s)
```

For reference about this behavior, see https://stackoverflow.com/a/18507871

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+#                    help='hidden units (default: [50, 20])')  #  https://docs.python.org/3/library/argparse.html#nargs
+parser.add_argument('--hidden_units', type=str, default='50 20', metavar='L',
+                    help='hidden units (default: "50 20")')
+parser.add_argument('--steps', type=int, default=1000, metavar='N',
```
Remove `metavar` from the parameter definition.

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+parser.add_argument('--hidden_units', type=str, default='50 20', metavar='L',
+                    help='hidden units (default: "50 20")')
+parser.add_argument('--steps', type=int, default=1000, metavar='N',
+                    help='steps (default: 1000)')
```
Nit: Instead of copying the default "1000" value, use `%(default)d`. Let's also make the help text a little more descriptive. The changed line will look like

```
help='The number of training steps (default: %(default)d)
```

For reference about this behavior, see https://stackoverflow.com/a/18507871

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+if __name__ == "__main__":
+    # The Estimator periodically generates "INFO" logs; make these logs visible.
+    tf.logging.set_verbosity(tf.logging.INFO)
+    #    tf.app.run(main=main)
```
Remove commented code

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
             saved_estimator_path = regressor.export_savedmodel(temp, receiver_fn).decode("utf-8")
             # Logging the saved model
-            tensorflow.log_saved_model(saved_model_dir=saved_estimator_path, signature_def_key="predict", artifact_path="model")
+            tensorflow.log_model( tf_saved_model_dir = saved_estimator_path , tf_signature_def_key = "predict" ,
```
Nit: Adjust spacing.

```
tensorflow.log_model(tf_saved_model_dir=saved_estimator_path , tf_signature_def_key="predict", artifact_path="model")
```

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+                tf_meta_graph_tags = [ args.tag_constants ] , artifact_path="model")
             # Reloading the model
-            pyfunc_model = pyfunc.load_pyfunc(saved_estimator_path)
+            pyfunc_model = pyfunc.load_pyfunc( os.path.join(mlflow.get_artifact_uri() , 'model' ) )  #  (saved_estimator_path)
```
Nit: Adjust spacing and remove `saved_estimator_path` comment.

```
pyfunc_model = pyfunc.load_pyfunc(os.path.join(mlflow.get_artifact_uri(), 'model'))
```

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+        tf.logging.set_verbosity(tf.logging.INFO)
+        # for key, value in vars(args).items():
+        #    mlflow.log_param(key, value)
+        hidden_units = [ int( s ) for s in args.hidden_units.split( ) ]  #  [50, 20]
```
Nit: Remove the comment at the end of this line that includes default values

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+        # for key, value in vars(args).items():
+        #    mlflow.log_param(key, value)
+        hidden_units = [ int( s ) for s in args.hidden_units.split( ) ]  #  [50, 20]
+        steps = args.steps  #  1000
```
Nit: Let's remove this line. We can refer to `args.steps` directly instead.

dbczumar commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
+        #    mlflow.log_param(key, value)
+        hidden_units = [ int( s ) for s in args.hidden_units.split( ) ]  #  [50, 20]
+        steps = args.steps  #  1000
+        print( '(hdnNts, stp) = ' , hidden_units , steps )
```
Let's include a description of the parameter information that we're logging and log these values with `mlflow.log_param`. Maybe we can replace this line with:

```
print("Hidden Units: {}".format(hidden_units))
print("Training Steps: {}".format(args.steps))
mlflow.log_param("hidden_units", hidden_units)
mlflow.log_param("training_steps", args.steps)
```

dbczumar commented on examples/tensorflow/MLproject:
```diff
@@ -0,0 +1,15 @@
+          python train_predict.py \
+            --hidden_units {hidden_units} \
+            --steps {steps} \
+            --tag_constants {tag_constants}
```
Remove `tag_constants` parameter

dbczumar commented on examples/tensorflow/MLproject:
```diff
@@ -0,0 +1,15 @@
+      tag_constants: {type: string, default: "serve"}
+    command: |
+          python train_predict.py \
+            --hidden_units {hidden_units} \
```
Let's use a hyphen instead of an underscore in the parameter name (`hidden_units` --> `hidden-units`)

dbczumar reviewed (changes requested):
@p2rk2h Thank you for the contribution! The general structure looks good. I left comments about the definition of certain `argparse` parameters, print/logging calls, and general code cleanliness. Let me know if you have any questions!

p2rk2h committed:
updates to pull request train_predict.py (#946)

MLproject updated (removed tags_constants; renamed to 'hidden-units')
train_predict.py updated per feedback comments; however, hardcoded mandatory tf_meta_graph_tags to avoid tensorflow error:
TypeError: log_model() missing 1 required positional argument: 'tf_meta_graph_tags=[tf.saved_model.tag_constants.SERVING]'

dbczumar committed:
Formatting, remove comment

dbczumar reviewed (approved):
@p2rk2h LGTM! Sorry for the delay. I made a couple of small formatting edits and will merge when tests pass.

p2rk2h commented on examples/tensorflow/train_predict.py:
```diff
@@ -2,53 +2,69 @@
             # Logging the saved model
-            tensorflow.log_saved_model(saved_model_dir=saved_estimator_path, signature_def_key="predict", artifact_path="model")
+            tensorflow.log_model( tf_saved_model_dir = saved_estimator_path , tf_signature_def_key = "predict" ,
+                tf_meta_graph_tags = [ args.tag_constants ] , artifact_path="model")
```
tensorflow.log_model now requires a new mandatory positional argument 'tf_meta_graph_tags' which I've hardcoded with [tf.saved_model.tag_constants.SERVING].  Would appreciate learning when would be appropriate to use [tf.saved_model.tag_constants.TRAINING] instead.
